### PR TITLE
Create OlympusINIReader.java

### PR DESCRIPTION
**Pull request by paulrichards321**
Reader for SlideScan.ini Olympus jpg file set by paulrichards321@gmail.com

### CODE REVIEW

**Comment by mtbc** [comment]
See also companion PR #2787 to add this reader to the list.

**Comment by paulrichards321** [comment]
I keep running into a java NullPointer exception on import. I have been debugging this for about a week with no luck. I can import the images if I do not assign resolutionCount > 1, but then I do not see the pyramid view. Attached is the log with the java runtime error. I have tried different version of bioformats, and grabbed yesterday's git version of the open microscopy source code and also tried 5.2.7 with no luck. I keep getting the same NullPointer error.

2017-03-08 16:32:28,615 19727      [      main] INFO   ormats.importer.cli.LoggingImportMonitor - FILESET_UPLOAD_END
2017-03-08 16:32:28,753 19865      [      main] INFO   ormats.importer.cli.LoggingImportMonitor - IMPORT_STARTED Logfile: 4000
2017-03-08 16:32:28,980 20092      [l.Client-1] ERROR     ome.formats.importer.cli.ErrorHandler - INTERNAL_EXCEPTION: /home/omero/SG09-599/SlideScan.ini
java.lang.RuntimeException: Failure response on import!
Category: ::omero::grid::ImportRequest
Name: import-request-failure
Parameters: {stacktrace=java.lang.NullPointerException
	at ome.formats.model.ChannelProcessor.process(ChannelProcessor.java:469)
	at ome.formats.OMEROMetadataStoreClient.postProcess(OMEROMetadataStoreClient.java:1752)
	at ome.services.blitz.repo.ManagedImportRequestI.importMetadata(ManagedImportRequestI.java:623)
	at ome.services.blitz.repo.ManagedImportRequestI.step(ManagedImportRequestI.java:501)
	at omero.cmd.HandleI.steps(HandleI.java:438)
	at omero.cmd.HandleI$1.doWork(HandleI.java:366)
	at omero.cmd.HandleI$1.doWork(HandleI.java:362)
	at sun.reflect.GeneratedMethodAccessor291.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
	at ome.services.util.Executor$Impl$Interceptor.invoke(Executor.java:562)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at ome.security.basic.EventHandler.invoke(EventHandler.java:154)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at org.springframework.orm.hibernate3.HibernateInterceptor.invoke(HibernateInterceptor.java:111)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:108)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at ome.tools.hibernate.ProxyCleanupFilter$Interceptor.invoke(ProxyCleanupFilter.java:249)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at ome.services.util.ServiceHandler.invoke(ServiceHandler.java:121)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202)
	at com.sun.proxy.$Proxy72.doWork(Unknown Source)
	at ome.services.util.Executor$Impl.execute(Executor.java:443)
	at omero.cmd.HandleI.run(HandleI.java:360)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at ome.services.util.Executor$Impl$1.call(Executor.java:484)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
, message=}

	at ome.formats.importer.ImportLibrary$ImportCallback.onFinished(ImportLibrary.java:665)
	at omero.cmd.CmdCallbackI.finished(CmdCallbackI.java:334)
	at omero.cmd._CmdCallbackDisp.___finished(_CmdCallbackDisp.java:118)
	at omero.cmd._CmdCallbackDisp.__dispatch(_CmdCallbackDisp.java:145)
	at IceInternal.Incoming.invoke(Incoming.java:222)
	at Ice.ConnectionI.invokeAll(ConnectionI.java:2482)
	at Ice.ConnectionI.dispatch(ConnectionI.java:1258)
	at Ice.ConnectionI.message(ConnectionI.java:1213)
	at IceInternal.ThreadPool.run(ThreadPool.java:321)
	at IceInternal.ThreadPool.access$300(ThreadPool.java:12)
	at IceInternal.ThreadPool$EventHandlerThread.run(ThreadPool.java:693)
	at java.lang.Thread.run(Thread.java:745)

java.lang.RuntimeException: Failure response on import!
Category: ::omero::grid::ImportRequest
Name: import-request-failure
Parameters: {stacktrace=java.lang.NullPointerException
	at ome.formats.model.ChannelProcessor.process(ChannelProcessor.java:469)
	at ome.formats.OMEROMetadataStoreClient.postProcess(OMEROMetadataStoreClient.java:1752)
	at ome.services.blitz.repo.ManagedImportRequestI.importMetadata(ManagedImportRequestI.java:623)
	at ome.services.blitz.repo.ManagedImportRequestI.step(ManagedImportRequestI.java:501)
	at omero.cmd.HandleI.steps(HandleI.java:438)
	at omero.cmd.HandleI$1.doWork(HandleI.java:366)
	at omero.cmd.HandleI$1.doWork(HandleI.java:362)
	at sun.reflect.GeneratedMethodAccessor291.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
	at ome.services.util.Executor$Impl$Interceptor.invoke(Executor.java:562)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at ome.security.basic.EventHandler.invoke(EventHandler.java:154)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at org.springframework.orm.hibernate3.HibernateInterceptor.invoke(HibernateInterceptor.java:111)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:108)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at ome.tools.hibernate.ProxyCleanupFilter$Interceptor.invoke(ProxyCleanupFilter.java:249)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at ome.services.util.ServiceHandler.invoke(ServiceHandler.java:121)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202)
	at com.sun.proxy.$Proxy72.doWork(Unknown Source)
	at ome.services.util.Executor$Impl.execute(Executor.java:443)
	at omero.cmd.HandleI.run(HandleI.java:360)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at ome.services.util.Executor$Impl$1.call(Executor.java:484)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
, message=}

	at ome.formats.importer.ImportLibrary$ImportCallback.onFinished(ImportLibrary.java:665) ~[blitz.jar:na]
	at omero.cmd.CmdCallbackI.finished(CmdCallbackI.java:334) [blitz.jar:na]
	at omero.cmd._CmdCallbackDisp.___finished(_CmdCallbackDisp.java:118) [blitz.jar:na]
	at omero.cmd._CmdCallbackDisp.__dispatch(_CmdCallbackDisp.java:145) [blitz.jar:na]
	at IceInternal.Incoming.invoke(Incoming.java:222) [ice.jar:na]
	at Ice.ConnectionI.invokeAll(ConnectionI.java:2482) [ice.jar:na]
	at Ice.ConnectionI.dispatch(ConnectionI.java:1258) [ice.jar:na]
	at Ice.ConnectionI.message(ConnectionI.java:1213) [ice.jar:na]
	at IceInternal.ThreadPool.run(ThreadPool.java:321) [ice.jar:na]
	at IceInternal.ThreadPool.access$300(ThreadPool.java:12) [ice.jar:na]
	at IceInternal.ThreadPool$EventHandlerThread.run(ThreadPool.java:693) [ice.jar:na]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_121]
2017-03-08 16:32:28,985 20097      [      main] ERROR        ome.formats.importer.ImportLibrary - Error on import
java.lang.Exception: Import failure
	at ome.formats.importer.ImportLibrary.importImage(ImportLibrary.java:560) ~[blitz.jar:na]
	at ome.formats.importer.ImportLibrary.importCandidates(ImportLibrary.java:289) ~[blitz.jar:na]
	at ome.formats.importer.cli.CommandLineImporter.start(CommandLineImporter.java:293) [blitz.jar:na]
	at ome.formats.importer.cli.CommandLineImporter.main(CommandLineImporter.java:950) [blitz.jar:na]
2017-03-08 16:32:28,985 20097      [      main] INFO         ome.formats.importer.ImportLibrary - Exiting on error

==> Summary
494 files uploaded, 0 filesets created, 0 images imported, 1 error in 0:00:17.810



David Gault has responded to my email to the mailing list with the following:

Hi Paul,

The work on the reader looks like it is coming along nicely. I was able to carry out some debugging of the exceptions you have been seeing. The first exception is due to a missing ExperimenterID which can be set as below at the same point which you are setting instrument and objective ID’s.

store.setExperimenterID(MetadataTools.createLSID("Experimenter", 0), 0);

Once that NullPointerException is resolved you will see a number of other exceptions relating to null values on the Pixels object for sizeC, sizeT, sizeZ etc. These largely relate to the handling of resolutions depending on if the flattenedResolutions is set to true or not. In this regard we had a numerous bits of feedback on various sections of code, would it be possible to upload the reader to GitHub? It might make it easier to provide comments on the individual sections or lines of code.

With Thanks,
David Gault

**Comment by sbesson** [comment]
Closing/reopening to re-trigger stale Travis build

**Comment by dgault** [comment]
Hi Paul,

The logic you have in place for setting the metadata and the resolutions is largely correct. In fact by default the reader should behave correctly. The issue is that OMERO uses an option not to flatten the resolutions (by default they are flattened), this changes the generated OME-XML but should be handled automatically. 

You can also test this option with the `showinf` tool by using the `-noflat` option (`showinf -noflat -nopix -omexml `filename will simply read and display the OME-XML which can be helpful).

So for the current reader we have the following generated by default. This will flatten the resolutions to give us 4 images and 4 sets of Pixel data which are all correctly populated as below:

```
<?xml version="1.0" encoding="UTF-8"?>
<OME xmlns="http://www.openmicroscopy.org/Schemas/OME/2016-06" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.openmicroscopy.org/Schemas/OME/2016-06 http://www.openmicroscopy.org/Schemas/OME/2016-06/ome.xsd">
   <Experimenter ID="Experimenter:0"/>
   <Instrument ID="Instrument:0">
      <Objective ID="Objective:0:0" NominalMagnification="40.0"/>
   </Instrument>
   <Image ID="Image:0" Name="1 (40x)">
      <InstrumentRef ID="Instrument:0"/>
      <ObjectiveSettings ID="Objective:0:0"/>
      <Pixels BigEndian="true" DimensionOrder="XYCZT" ID="Pixels:0" Interleaved="true" PhysicalSizeX="2.0691489361702127" PhysicalSizeXUnit="µm" PhysicalSizeY="2.39375" PhysicalSizeYUnit="µm" SignificantBits="8" SizeC="3" SizeT="1" SizeX="15040" SizeY="15360" SizeZ="1" Type="uint8">
         <Channel ID="Channel:0:0" SamplesPerPixel="3">
            <LightPath/>
         </Channel>
         <MetadataOnly/>
      </Pixels>
   </Image>
   <Image ID="Image:1" Name="2 (20x)">
      <InstrumentRef ID="Instrument:0"/>
      <ObjectiveSettings ID="Objective:0:0"/>
      <Pixels BigEndian="true" DimensionOrder="XYCZT" ID="Pixels:1" Interleaved="true" PhysicalSizeX="8.27659574468085" PhysicalSizeXUnit="µm" PhysicalSizeY="9.575" PhysicalSizeYUnit="µm" SignificantBits="8" SizeC="3" SizeT="1" SizeX="3760" SizeY="3840" SizeZ="1" Type="uint8">
         <Channel ID="Channel:1:0" SamplesPerPixel="3">
            <LightPath/>
         </Channel>
         <MetadataOnly/>
      </Pixels>
   </Image>
   <Image ID="Image:2" Name="3 (5x)">
      <InstrumentRef ID="Instrument:0"/>
      <ObjectiveSettings ID="Objective:0:0"/>
      <Pixels BigEndian="true" DimensionOrder="XYCZT" ID="Pixels:2" Interleaved="true" PhysicalSizeX="66.23005319148936" PhysicalSizeXUnit="µm" PhysicalSizeY="76.64166666666667" PhysicalSizeYUnit="µm" SignificantBits="8" SizeC="3" SizeT="1" SizeX="469" SizeY="479" SizeZ="1" Type="uint8">
         <Channel ID="Channel:2:0" SamplesPerPixel="3">
            <LightPath/>
         </Channel>
         <MetadataOnly/>
      </Pixels>
   </Image>
   <Image ID="Image:3" Name="4 (1.25x)">
      <InstrumentRef ID="Instrument:0"/>
      <ObjectiveSettings ID="Objective:0:0"/>
      <Pixels BigEndian="true" DimensionOrder="XYCZT" ID="Pixels:3" Interleaved="true" PhysicalSizeX="264.92021276595744" PhysicalSizeXUnit="µm" PhysicalSizeY="306.56666666666666" PhysicalSizeYUnit="µm" SignificantBits="8" SizeC="3" SizeT="1" SizeX="117" SizeY="119" SizeZ="1" Type="uint8">
         <Channel ID="Channel:3:0" SamplesPerPixel="3">
            <LightPath/>
         </Channel>
         <MetadataOnly/>
      </Pixels>
   </Image>
</OME>
```

What we are expecting to see with the `-noflat` option or through OMERO is that we have a single image with multiple resolutions and the XML generated should look as follows:

```
<?xml version="1.0" encoding="UTF-8"?>
<OME xmlns="http://www.openmicroscopy.org/Schemas/OME/2016-06" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.openmicroscopy.org/Schemas/OME/2016-06 http://www.openmicroscopy.org/Schemas/OME/2016-06/ome.xsd">
   <Experimenter ID="Experimenter:0"/>
   <Instrument ID="Instrument:0">
      <Objective ID="Objective:0:0" NominalMagnification="40.0"/>
   </Instrument>
   <Image ID="Image:0" Name="1 (40x)">
      <InstrumentRef ID="Instrument:0"/>
      <ObjectiveSettings ID="Objective:0:0"/>
      <Pixels BigEndian="true" DimensionOrder="XYCZT" ID="Pixels:0" Interleaved="true" PhysicalSizeX="2.0691489361702127" PhysicalSizeXUnit="µm" PhysicalSizeY="2.39375" PhysicalSizeYUnit="µm" SignificantBits="8" SizeC="3" SizeT="1" SizeX="15040" SizeY="15360" SizeZ="1" Type="uint8">
         <Channel ID="Channel:0:0" SamplesPerPixel="3">
            <LightPath/>
         </Channel>
         <MetadataOnly/>
      </Pixels>
   </Image>
</OME>
```

What actually happens is that 4 images and 4 sets of Pixel objects are generated. Though because we are now only expecting a single image the Pixel data is not fully populated and throws the null pointer exceptions you are seeing. The generated XML is as follows:

```
<?xml version="1.0" encoding="UTF-8"?>
<OME xmlns="http://www.openmicroscopy.org/Schemas/OME/2016-06" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.openmicroscopy.org/Schemas/OME/2016-06 http://www.openmicroscopy.org/Schemas/OME/2016-06/ome.xsd">
   <Experimenter ID="Experimenter:0"/>
   <Instrument ID="Instrument:0">
      <Objective ID="Objective:0:0" NominalMagnification="40.0"/>
   </Instrument>
   <Image ID="Image:0" Name="1 (40x)">
      <InstrumentRef ID="Instrument:0"/>
      <ObjectiveSettings ID="Objective:0:0"/>
      <Pixels BigEndian="true" DimensionOrder="XYCZT" ID="Pixels:0" Interleaved="true" PhysicalSizeX="2.0691489361702127" PhysicalSizeXUnit="µm" PhysicalSizeY="2.39375" PhysicalSizeYUnit="µm" SignificantBits="8" SizeC="3" SizeT="1" SizeX="15040" SizeY="15360" SizeZ="1" Type="uint8">
         <Channel ID="Channel:0:0" SamplesPerPixel="3">
            <LightPath/>
         </Channel>
         <MetadataOnly/>
      </Pixels>
   </Image>
   <Image Name="2 (20x)">
      <InstrumentRef ID="Instrument:0"/>
      <ObjectiveSettings ID="Objective:0:0"/>
      <Pixels PhysicalSizeX="8.27659574468085" PhysicalSizeXUnit="µm" PhysicalSizeY="9.575" PhysicalSizeYUnit="µm"/>
   </Image>
   <Image Name="3 (5x)">
      <InstrumentRef ID="Instrument:0"/>
      <ObjectiveSettings ID="Objective:0:0"/>
      <Pixels PhysicalSizeX="66.23005319148936" PhysicalSizeXUnit="µm" PhysicalSizeY="76.64166666666667" PhysicalSizeYUnit="µm"/>
   </Image>
   <Image Name="4 (1.25x)">
      <InstrumentRef ID="Instrument:0"/>
      <ObjectiveSettings ID="Objective:0:0"/>
      <Pixels PhysicalSizeX="264.92021276595744" PhysicalSizeXUnit="µm" PhysicalSizeY="306.56666666666666" PhysicalSizeYUnit="µm"/>
   </Image>
</OME>

```

The root cause of this is because the Image and Pixel metadata is generated based on the number of metadata cores each time. This is correct in the scenario in which we flatten resolutions but leaves us with extra objects when we don't. The reason they appear partially populated is because the `MetadataTools.populatePixels(store, this);` adjusts its expected series count based on the resolutionCount, thus it only populates the first set of Pixel data. 

The solution to this is that we can use getSeriesCount() when setting the Pixel data and the Image Name attributes. This will adjust automatically adjust based on the `resolutionCount` and the `flattenedResolutions` variables.

File: components/formats-gpl/src/loci/formats/in/OlympusINIReader.java [hunk 1]
```diff
@@ -0,0 +1,950 @@
...
+      String objectiveID = MetadataTools.createLSID("Objective", 0, 0);
+      store.setObjectiveID(objectiveID, 0, 0);
+      store.setObjectiveNominalMagnification(magnification, 0, 0);
+
+      store.setExperimenterID(MetadataTools.createLSID("Experimenter", 0), 0);
+
+
+      totalCores = 0;
+      for (int fileNum=0; fileNum < 4; fileNum++) 
+      {
```
Comment (dgault) [inline components/formats-gpl/src/loci/formats/in/OlympusINIReader.java]: So in this case we only want to populate Pixel metadata based on the expected length of the series. So you could use:
`for (int fileNum=0; fileNum < getSeriesCount(); fileNum++) `

File: components/formats-gpl/src/loci/formats/in/OlympusINIReader.java [hunk 2]
```diff
@@ -0,0 +1,950 @@
...
+        }
+        totalCores++;
+      }
+    }
+    MetadataStore store = makeFilterMetadata();
+    MetadataTools.populatePixels(store, this);
+
+    String path = new Location(currentId).getAbsoluteFile().getName();
+
+    if (totalLevels > 3)
```
Comment (dgault) [inline components/formats-gpl/src/loci/formats/in/OlympusINIReader.java]: For this series of if statements settings the ImageName, it is always creating 4 image objects regardless of if the resolutions are being flattened or not. Here you will also need to use the getSeriesCount() to only set ImageName on Images upto the expected series length.

**Comment by sbesson** [comment]
Note the file detection logic seems to conflict with other file formats based on `.ini` files (see this [log](https://ci.openmicroscopy.org/job/BIOFORMATS-DEV-breaking-repository-subset/231) for SVS files. Closing this PR and the companion `readers.txt` to prevent false positives in the integration builds.

**Pull request closed without merging**